Unit tests

Edgarware opened the pull request:
*AKA: Edgarware performs pytest crimes*

Implements unit testing of all models using Github Actions. It'll generate build artifacts for each tested python version with XML & HTML test logs and coverage reports.

Notes:
* You might need to update github notification settings so that it doesnt ping you every time a build happens :P
* ~The github action tests python versions 3.6+, which all seem to work fine? I'm not sure why we're requiring 3.8 right now but it seems like we can drop that requirement without issue.~
* ~There's 2 commits before the Big One that make debugging a little nicer and fix some weirdness with JSON exports. I can throw these into a separate PR if that's better for you.~
* Most of the tests are just "Does this field exist and is it the right type according to the Blaseball API". Adding new field tests is pretty easy and hopefully they wont remove fields (🤞) but I can also just remove those tests if it's too much of a maintenance burden.
* ~I currently have test stubs to verify that we error correctly on the load functions if you pass garbage data, but we probably need to standardize how we're going to error in these cases.~
* This doesn't test the eventsteam stuff, mostly because I'm not sure how to... do that.

jmaliksi commented:
oh fun

jmaliksi commented:
it might take me a while to review this

Edgarware commented:
lol yeah no rush. This is mostly just so that I can get it off my computer.

jmaliksi commented on tests/test_player.py:
```diff
@@ -0,0 +1,214 @@
+"""
+Unit Tests for Player Model
+
+These tests may fail at any time due to the whims of the Blaseball API. See also: Blaseball Falsehoods
```
let's make this not a thing and just use vcr https://vcrpy.readthedocs.io/en/latest/

jmaliksi commented on tests/test_election.py:
```diff
@@ -0,0 +1,133 @@
+from tests.test_base import base_test
+
+
+def pytest_generate_tests(metafunc):
```
while neat, I think using `pytest_generate_tests` is going to add a bunch of maintenance overhead given all the introspection, and the fact that it dynamically parametrizes using using the live API adds instability and uncertainty. Even after using VCR to lock the HTTP response, it's still going to be difficult to debug the input if something weird happens. Also `pytest_generate_tests` is like really poorly documented.

My preference is to define static fixtures in `conftest`. You can even use the same instantiation logic so it's not a complete rewrite. Even though that doesn't fully solve the second point, at the very least it makes the fixtures reusable and a bit easier to copy/edit as the new cases arise going forward.

One other thing that this shotgun approach is causing a 10 minute test run. This library is not nearly large or complex enough to warrant a 10 minute test run. Part of this is definitely the HTTP calls, but also I suspect that many of the test cases that get automagically loaded aren't testing any new code paths.

Edgarware replied:
I agree, particularly after looking into VCR more. Implementing it with the existing collection steps is... not fun. 

The main reason I built it like this is to verify that the load functions return what we expect, which has already found a number of issues (one of which I just found yesterday with Player.load_by_gameday) but you're right that we really only need one for each load function and that can probably just be in a separate fixture-ized test function for each load type.

And I didn't actually know about `conftest.py` until doing some digging this morning. pytest sure seems to like having features defined by poorly defined file/function names...

jmaliksi commented on tests/test_game.py:
```diff
@@ -0,0 +1,249 @@
+    """
+    Test that players & player fields do not error
+    """
+    assert isinstance(game.away_pitcher, (Player, type(None)))
```
this is a pretty good example of where explicit fixtures and parametrization would make this more robust. `None` is not a correct pitcher value for basically any valid game in the past, and not `None` isn't necessarily correct for any game in the future.

Edgarware replied:
Your reasoning is correct but hilariously enough `game.awayPitcher` is actually None in some cases https://www.blaseball.com/database/gameById/50a6fc92-acdd-4999-a1c1-1e5e1eedc685

jmaliksi commented:
> The github action tests python versions 3.6+, which all seem to work fine? I'm not sure why we're requiring 3.8 right now but it seems like we can drop that requirement without issue.
I think it was something to do with the even stream, so let's leave that for a separate look

> There's 2 commits before the Big One that make debugging a little nicer and fix some weirdness with JSON exports. I can throw these into a separate PR if that's better for you.
Fine to keep it here for now, but if it takes longer than expected to integrate VCR or to figure out how we want to properly parametrize things, then I'd say kick it out so it can live on main

> Most of the tests are just "Does this field exist and is it the right type according to the Blaseball API". Adding new field tests is pretty easy and hopefully they wont remove fields (🤞) but I can also just remove those tests if it's too much of a maintenance burden.
this is probably fine for now.

> I currently have test stubs to verify that we error correctly on the load functions if you pass garbage data, but we probably need to standardize how we're going to error in these cases.
i feel like `ValueError` is fine

> This doesn't test the eventsteam stuff, mostly because I'm not sure how to... do that.
let's definitely save that for another PR

Edgarware commented on tests/test_election.py:
```diff
@@ -0,0 +1,133 @@
+    #    DecreeResult.load_one("00000000-0000-0000-0000-000000000000")
+    #with pytest.raises(ValueError):
+    #    TidingResult.load_one("00000000-0000-0000-0000-000000000000")
+    pass
```
Same as the other test_error functions, commented them out while we figured out how to standardize errors. But you answered my question about that so I'll update it

Edgarware commented:
> Fine to keep it here for now, but if it takes longer than expected to integrate VCR or to figure out how we want to properly parametrize things, then I'd say kick it out so it can live on main

I'll probably move these to a separate PR then, since I have some other fixes to roll in anyway.

Edgarware commented:
OK, this should be better. This depends on #77, the commits of which are in here so it passes tests.

* Uses VCR to record and play-back network calls. This produces a lot of cassettes since we have to deal with lazy-load checks in each model test suite. The naming convention that `pytest-vcr` uses is pretty straightforward though.
* Got rid of the `pytest_generate_tests` stuff, replaced with regular `fixtures` in `conftest.py`.  In addition the fixtures are doing a lot fewer tests than before, more of which are loaded directly from local JSON files to maximize test coverage.
* Implemented error check testing. There's some weirdness with some of the loads with respect to when they raise Errors and when they don't, but that can be dealt with in a future PR.
* Only testing Python 3.8 & 3.9 for now.

With these changes each test run went from taking ~2 minutes to ~3 seconds, so a good speedup for sure!

jmaliksi commented on blaseball_mike/chronicler.py:
```diff
@@ -6,14 +6,30 @@
 
+
+def _check_response(response):
+    if not response.ok:
```
let's use `raise_for_status`

jmaliksi commented on blaseball_mike/database.py:
```diff
@@ -4,16 +4,29 @@
 BASE_GITHUB = 'https://raw.githubusercontent.com/xSke/blaseball-site-files/main/data'
 
 
+def _check_response(response):
```
we can probably kick this out into the `session` file or something for all common http related helpers

jmaliksi commented:
i have never used gitattributes before but i wonder if this will work to automatically collapse the yaml files in the diff for easier review? https://stackoverflow.com/a/32511147

Edgarware commented:
> i have never used gitattributes before but i wonder if this will work to automatically collapse the yaml files in the diff for easier review? https://stackoverflow.com/a/32511147

Hmmm this works in the Github Desktop app but not on the site?

jmaliksi commented:
aw dang, ok. if you can figure out in under 5 minutes how to tell github
that a file is generated, then would be useful. I'm pretty sure it's
possible but I never looked into it myself. Although we also want to make
sure that the generated files aren't hidden from review cuz sometimes it's
good to scrub re-records for secrets or redundant data (indicating possible
leak)

On Mon, Feb 15, 2021 at 11:32 AM Ryan Littleton <notifications@github.com>
wrote:

> i have never used gitattributes before but i wonder if this will work to
> automatically collapse the yaml files in the diff for easier review?
> https://stackoverflow.com/a/32511147
>
> Hmmm this works in the Github Desktop app but not on the site?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/jmaliksi/blaseball-mike/pull/73#issuecomment-779332616>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAYCOCFVL6SVEJQB4SG3FYLS7FEDFANCNFSM4XNS43YQ>
> .
>

Edgarware commented:
> 
> aw dang, ok. if you can figure out in under 5 minutes how to tell github that a file is generated, then would be useful. I'm pretty sure it's possible but I never looked into it myself. Although we also want to make sure that the generated files aren't hidden from review cuz sometimes it's good to scrub re-records for secrets or redundant data (indicating possible leak)

Figured it out. Turns out Github needs a [special code word].(https://github.com/github/linguist/issues/2206#issuecomment-339272910)

jmaliksi commented on tests/test_player.py:
```diff
@@ -0,0 +1,303 @@
+                    assert player.get_vibe(day) == pytest.approx(player.vibes[day-1])
+
+    @pytest.mark.vcr
+    def test_blood_coffee_bounded(self):
```
probably don't need a json file for this, just define the json in two separate tests. One thing to keep in mind is if you have a single test doing a loop of cases or a slew of assertions, you're probably undermining pytest and should convert to parametrization and/or multiple test cases.

A loop is suboptimal for two major reasons:
1) if one of the cases early in the loop fails, problems in later cases will not arise until you fix them all in series
2) mixing cases muddles the stack trace and isn't clear which input caused the problem

Edgarware replied:
So a fun problem with parametrizing this one specifically:
* In order to accurately test this, we need the error case, so we get "Coffee?" or "Blood?" as our expected result
* This test needs VCR, since the blood/coffee calls are network calls
* pytest-vcr is smart and will include the parameter name in the cassette file name
* But then Windows dies because we have a question mark in a filename....

Edgarware replied:
OK, fixed this one. Turns out `pytest-vcr` is outdated and the creator made `pytest-recording` which fixes this bug.

jmaliksi commented on tests/test_player.py:
```diff
@@ -0,0 +1,303 @@
+        assert isinstance(test_value, str)
+        assert test_value[-9:] == "undefined"
+
+    def test_player_make_random(self):
```
separate these into two different test cases. consider moving Blintz from json to a fixture in conftest

jmaliksi commented on tests/test_player.py:
```diff
@@ -0,0 +1,303 @@
+        """
+        Test that soulscream equation produces correct results
+        """
+        with open(f"{TEST_DATA_DIR}/player/screams.json", "r") as fp:
```
same comment as payouts in that parametrization of the test and removal of the json imo would be cleaner. It might be annoying to define all the attributes in the parametrization decorator, so going back to another earlier comment in conftest, if you define each player as their own fixture in conftest, you could just reuse those (i forgot how to use a list of fixtures in parametrization, though)

jmaliksi commented on tests/test_player.py:
```diff
@@ -0,0 +1,303 @@
+        """
+        Test that star equations produce correct results
+        """
+        with open(f"{TEST_DATA_DIR}/player/stars.json", "r") as fp:
```
imo would be better to have three separate, explicit test cases. see comment for `test_blood_coffee_bounded` for further reasoning

jmaliksi commented on tests/test_player.py:
```diff
@@ -0,0 +1,303 @@
+        """
+        Test that vibe equation produces correct results
+        """
+        with open(f"{TEST_DATA_DIR}/player/vibes.json", "r") as fp:
```
having the expected vibes defined in json makes sense here, but again, I'd consider having multiple test cases rather than a loop. though I'm not even sure if this needs the multiple test cases--the logic is the same regardless of what attribute is high or low, so it's the same amount of coverage for each

jmaliksi commented on tests/test_team.py:
```diff
@@ -0,0 +1,119 @@
+            assert key == team.id
+
+    @pytest.mark.vcr
+    def test_load_at_time(self):
```
make these separate test cases

I was about to make the same comment a bunch of times, so general guidance: if a test loads multiple objects, separate them into different tests that are explicitly checking for a specific output (ie in the simplest cases you have everywhere, one test for valid, one test for invalid)

like sure, unfortunately it's more boilerplate, but it's easier to debug in the future and will be better at pinpointing any issues that arise

jmaliksi commented on tests/test_team.py:
```diff
@@ -0,0 +1,119 @@
+    @pytest.mark.vcr
+    def test_load_all(self):
+        teams = Team.load_all()
+        assert isinstance(teams, dict)
```
assert size of dictionary as an empty return will pass this test

Edgarware commented on tests/conftest.py:
```diff
@@ -0,0 +1,277 @@
+
+@pytest.fixture(scope="module")
+@vcr.use_cassette(f'{CASSETTE_DIR}/Fixture.games.yaml')
+def games():
```
Do you want to have the static ones (like crowvertime) load the `dict` from a json or do you want it defined in `conftest.py`? I worry having it hard-coded will just bloat the config file.

The reason I used the singular JSON for each was to simplify adding new test cases. You only have to append it to one file rather than adding a JSON file and adding a fixture and adding that fixture to the group fixture.

jmaliksi replied:
imo a bloated conftest is fine (that's a problem solved by the interpreter, so let's not reinvent the wheel), so yeah, static ones can be moved in there.

if the json file is removed, adding new cases is just two touch points (new fixture, append to group fixture), though as mentioned in other comments, it would be best if we could take advantage of parametrization at the test level, which solves the group fixture touch point. 

My general reasoning for preferring a bunch of small fixtures is that if a new test case arises, that really should only add one test for that case, not n+1 tests for all the existing prior cases. Those prior cases should, in theory, already be covered. 

Like for example, let's say there's a bug found in calculating payouts on sun2 games--therefore we should only test payouts on a sun2 game when making the fix. We don't need to test initialization as there's no special case, we don't need to test player mods as weather loading and player loading are independent, etc.

Edgarware replied:
New update has split the stuff out. The problem is you cant call a fixture inside a fixture, so it's currently setup that the *group* is a fixture and the rest are just standard functions. This isn't ideal, since it means you cant explicitly call specific fixtures for one-off tests. I also dont think the group can be parameterized if it's a fixture, unless we do the `pytest_generate_tests` stuff again...

jmaliksi replied:
wait you can't call a fixture inside a fixture? i feel like i have done this before....

what i was hoping for was that the for loop could be taken out of the individual test cases and just rely on the parametrization scaffolding, so now we're kind of in a worst-of-both-worlds situation.

some cursory poking around, I found a thing that I've never used. I wonder if's any good? https://pypi.org/project/pytest-lazy-fixture/

Edgarware replied:
I *think* I solved it. So if we define the singular test functions as fixtures, then perform parametrization *on the group fixture*, we can do this:

```python
@pytest.fixture(scope="module", params=['game_s7d95', 'game_s2d99_chronicler', 'game_s9d1_crovertime', 'game_s2d10_none_awaypitcher', 'game_coffee_cup_d11'])
def game(request):
    return request.getfixturevalue(request.param)
```

So any test function that calls `game` as an argument will be automatically parametrized with the tests in the `params` list.

jmaliksi replied:
ok yeah, i think this works pretty well, even though it looks like a mess. as long as the mess is contained to conftest, I think that's a win.

Edgarware replied:
OK, I've done enough of them that the test dont error. I'll go back and finish up the last of them after dinner.

Is there any interest in moving these out of conftest.py? The tests are grouped by class so we could just put the fixtures in with their respective tests.

jmaliksi replied:
yeah separating them out is probably fine to keep it svelt

jmaliksi commented on tests/README.md:
````diff
@@ -0,0 +1,28 @@
+```
+---
+### I'm getting 'vcr.errors.CannotOverwriteExistingCassetteException' errors...
+We use [vcrpy](https://vcrpy.readthedocs.io/en/latest/index.html) to record network calls
````
update link?

Edgarware replied:
That's the generic `vcr` documentation, but I can add a link to the `pytest-recording` github. Their documentation is a lot worse though.

jmaliksi commented on tests/test_election.py:
```diff
@@ -0,0 +1,260 @@
+    def test_load_by_season(self):
+        result = ElectionResult.load_by_season(1)
+        assert isinstance(result, ElectionResult)
+        # TODO: oops whoops uh oh
```
uh?

Edgarware replied:
Yet another case of us forgetting to 1-index a season. But fixing it'd be a breaking change and I'm not sure it's worth doing another Major release since we just did one last week.

jmaliksi replied:
i'd consider fixing an off by one to be a bug fix tbh, so go ahead and fix it

Edgarware replied:
Wasnt sure on this one, but it's fixed. Realistically it's impossible for someone to get the `ElectionResults` without already knowing what season they're looking up anyway.

jmaliksi reviewed (approved):
this is great!
